gateway-ha/src/main/java/io/trino/gateway/ha/router/GatewayCookie.java

```diff
@@ -0,0 +1,241 @@
+    {
+        this.deletePaths = deletePaths;
+    }
+
+    private String computeSignature()
+    {
+        return Hashing.hmacSha256(GatewayCookieConfiguration.getCookieSigningKey())
+                .hashString(Integer.toString(hashCode()), UTF_8)
+                .toString();
+    }
+
+    @Override
+    public boolean equals(Object obj)
+    {
+        if (this == obj) {
+            return true;
+        }
+
+        if (!(obj instanceof GatewayCookie)) {
+            return false;
+        }
+
+        GatewayCookie other = (GatewayCookie) obj;
+        if (other.ts.equals(ts)
+                && other.name.equals(name)
+                && other.routingPaths.equals(routingPaths)
+                && other.deletePaths.equals(deletePaths)
+                && other.priority == priority
+                && other.lifetime.equals(lifetime)
+                && ((payload != null && payload.equals(other.payload)) || (payload == other.payload))) {
+            return true;
+        }
+        return false;
+    }
+
+    @Override
+    public int hashCode()
+    {
+        int result = 7;
+        result = 37 * result + ts.hashCode();
+        result = 37 * result + name.hashCode();
+        result = payload != null ? 37 * result + payload.hashCode() : result;
+        result = 37 * result + routingPaths.hashCode();
+        result = 37 * result + deletePaths.hashCode();
+        result = 37 * result + priority;
+        result = 37 * result + lifetime.hashCode();
+        result = backend != null ? 37 * result + backend.hashCode() : result;
+
+        return result;
+    }
+
+    @Override
+    public int compareTo(GatewayCookie o)
+    {
+        int priorityDelta = priority - o.getPriority();
+        return priorityDelta != 0 ? priorityDelta : (int) (ts - o.getTs());
+    }
+
+    public Cookie toCookie()
+    {
+        Cookie cookie = new Cookie(name, Base64.getUrlEncoder().encodeToString(CODEC.toJson(this).getBytes(UTF_8)));
+        cookie.setMaxAge((int) lifetime.toMillis() / 1000);
+        return cookie;
+    }
+
+    public static GatewayCookie fromCookie(Cookie cookie)
+    {
+        return GatewayCookie.CODEC.fromJson(Base64.getUrlDecoder().decode(cookie.getValue()));
+    }
+
+    public boolean matchesRoutingPath(String path)
+    {
+        for (String cookiePath : routingPaths) {
+            if (!deletePaths.contains(cookiePath) && path.startsWith(cookiePath)) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    public boolean matchesDeletePath(String path)
+    {
+        return deletePaths.contains(path);
+    }
+
+    public boolean isValid()
+    {
+        if (System.currentTimeMillis() > ts + lifetime.toMillis()) {
+            return false;
+        }
+
+        if (GatewayCookieConfiguration.isValidate()) {
+            if (isNullOrEmpty(signature)) {
+                return false;
+            }
+            return signature.equals(computeSignature());
+        }
```

Comment:
Why not just always sign and verify the cookie?

Reply:
I wouldn't mind removing this option. I figured some people may complain about overhead.

Reply:
@Chaho12 wdyt? Any complaints about always validating? There is the cost of computing a cryptographic hash, which isn't free but probably a small fraction of the total cost per request.

Reply:
Do you know how small the fraction is?

Reply:
It should be roughly similar to the cost of initiating a TLS handshake

Reply:
Then i think it is small enough to be ignored :)